nixos/modules/services/web-apps/moin.nix

```diff
@@ -0,0 +1,291 @@
+          {
+            description = "MoinMoin wiki ${wikiIdent} - gunicorn process";
+            wantedBy = [ "multi-user.target" ];
+            after = [ "network.target" ];
+            restartIfChanged = true;
+            restartTriggers = [ (wikiConfigFile wikiIdent wiki) ];
+
+            environment = let
+              penv = python.buildEnv.override {
+                # setuptools: https://github.com/benoitc/gunicorn/issues/1716
+                extraLibs = [ python.pkgs.gevent python.pkgs.setuptools pkg ];
+              };
+            in {
+              PYTHONPATH= "${dataDir}/${wikiIdent}/config:${penv}/${python.sitePackages}";
+            };
+
+            serviceConfig = {
+              Type = "simple";
+              User = user;
+              Group = group;
+              WorkingDirectory = "${dataDir}/${wikiIdent}";
+              ExecStart = ''${python.pkgs.gunicorn}/bin/gunicorn moin_wsgi \
+                --name gunicorn-${wikiIdent} \
+                --workers ${toString cfg.gunicorn.workers} \
+                --worker-class gevent \
+                --bind unix:/run/moin/${wikiIdent}/gunicorn.sock
+              '';
+
+              StateDirectory = "moin/${wikiIdent}";
+              StateDirectoryMode = "0550";
+              RuntimeDirectory = "moin/${wikiIdent}";
+              RuntimeDirectoryMode = "0750";
+
+              NoNewPrivileges = true;
+              ProtectSystem = "strict";
+              ProtectHome = true;
+              PrivateTmp = true;
+              PrivateDevices = true;
+              PrivateNetwork = true;
+              ProtectKernelTunables = true;
+              ProtectKernelModules = true;
+              ProtectControlGroups = true;
+              RestrictAddressFamilies = [ "AF_UNIX" "AF_INET" "AF_INET6" ];
+              RestrictNamespaces = true;
+              LockPersonality = true;
+              MemoryDenyWriteExecute = true;
+              RestrictRealtime = true;
+            };
+          }
+      ));
+
+    services.nginx = mkIf usingNginx {
+      enable = true;
+      virtualHosts = flip mapAttrs' cfg.wikis (name: w: nameValuePair w.webHost {
+        forceSSL = mkDefault true;
+        enableACME = mkDefault true;
+        locations."${w.webLocation}" = {
+          extraConfig = ''
+            proxy_set_header Host $host;
+            proxy_set_header X-Real-IP $remote_addr;
+            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+            proxy_set_header X-Forwarded-Proto $scheme;
+            proxy_set_header X-Forwarded-Host $host;
+            proxy_set_header X-Forwarded-Server $host;
+
+            proxy_pass http://unix:/run/moin/${name}/gunicorn.sock;
+          '';
+        };
+      });
+    };
+
+    systemd.tmpfiles.rules = [
+      "d  /run/moin            0750 ${user} ${group} - -"
+      "d  ${dataDir}           0550 ${user} ${group} - -"
+    ]
+    ++ (concatLists (flip mapAttrsToList cfg.wikis (wikiIdent: wiki: [
```

Comment:
Is there any concept of 'migrations' in `moinmoin` which might have some problems introduced during version upgrades here (particularly the `C` entries in your `tmpfiles`)? While I do like the declarative nature of `tmpfiles` I have concerns that `tmpfiles` may not be flexible enough **if** upgrading between versions of `moinmoin` requires special handling.

I'm not familiar with `moinmoin` as a sysadmin, so if my question is of no concern please disregard.

Reply:
The `data` dir is supposed to be migrated to newer version using
`moin --config-dir=/path/to/config_dir --wiki-url=wiki.example.org/ migration data`.
Not sure whether we should do this automatically, however the `C` should be fine here for manual migrations.

Regarding the `underlay` dir the docs suggest to replace it with the one from the newer version. We can make `preStart` always replace it at the cost of slower startup. Symlinking doesn't work unfortunately.

References:
* https://github.com/moinwiki/moin-1.9/blob/master/docs/README.migration
* https://master19.moinmo.in/HelpOnUpdating